File: typescript/src/logger.ts
/** Pluggable logger interface — pino's Logger satisfies this directly. */
export interface ChaosLogger {
  debug(obj: Record<string, unknown>, msg: string): void;
  info(obj: Record<string, unknown>, msg: string): void;
  info(msg: string): void;
  warn(obj: Record<string, unknown>, msg: string): void;
  error(obj: Record<string, unknown>, msg: string): void;
}

export type ChaosLoggerFactory = (name: string) => ChaosLogger;

let loggerFactory: ChaosLoggerFactory = createConsoleLogger;

/** Create a default console-based logger. */
export function createConsoleLogger(name: string): ChaosLogger {
  const prefix = `[${name}]`;
  return {
    debug(objOrMsg: Record<string, unknown> | string, msg?: string): void {
      if (typeof objOrMsg === 'string') {
        console.debug(prefix, objOrMsg);
      } else {
        console.debug(prefix, msg, objOrMsg);
      }
    },
    info(objOrMsg: Record<string, unknown> | string, msg?: string): void {
      if (typeof objOrMsg === 'string') {
        console.info(prefix, objOrMsg);
      } else {
        console.info(prefix, msg, objOrMsg);
      }
    },
    warn(objOrMsg: Record<string, unknown> | string, msg?: string): void {
      if (typeof objOrMsg === 'string') {
        console.warn(prefix, objOrMsg);
      } else {
        console.warn(prefix, msg, objOrMsg);
      }
    },
    error(objOrMsg: Record<string, unknown> | string, msg?: string): void {
      if (typeof objOrMsg === 'string') {
        console.error(prefix, objOrMsg);
      } else {
        console.error(prefix, msg, objOrMsg);
      }
    },
  } as ChaosLogger;
}

/** Set the global logger factory. Call once at startup. */
export function configureChaosLogger(factory: ChaosLoggerFactory): void {
  loggerFactory = factory;
}

/** Internal — used by all library modules. */
export function getLogger(name: string): ChaosLogger {
  return loggerFactory(name);
}

File: python/src/mcp_chaos_monkey/logger.py
from __future__ import annotations

import logging
from collections.abc import Callable
from typing import Any, Protocol, runtime_checkable


@runtime_checkable
class ChaosLogger(Protocol):
    """Pluggable logger interface — stdlib logging.Logger satisfies this directly."""

    def debug(self, msg: str, *args: Any, **kwargs: Any) -> None: ...
    def info(self, msg: str, *args: Any, **kwargs: Any) -> None: ...
    def warning(self, msg: str, *args: Any, **kwargs: Any) -> None: ...
    def error(self, msg: str, *args: Any, **kwargs: Any) -> None: ...


ChaosLoggerFactory = Callable[[str], ChaosLogger]


def create_default_logger(name: str) -> ChaosLogger:
    """Create a default logger using stdlib logging."""
    return logging.getLogger(f"mcp_chaos_monkey.{name}")


_logger_factory: ChaosLoggerFactory = create_default_logger


def configure_chaos_logger(factory: ChaosLoggerFactory) -> None:
    """Set the global logger factory. Call once at startup."""
    global _logger_factory
    _logger_factory = factory


class _LazyLogger:
    """Proxy that delegates to the current logger factory on each call.

    This ensures that calling ``configure_chaos_logger`` after modules have
    already been imported still takes effect.
    """

    __slots__ = ("_name",)

    def __init__(self, name: str) -> None:
        self._name = name

    def _get(self) -> ChaosLogger:
        return _logger_factory(self._name)

    def debug(self, msg: str, *args: Any, **kwargs: Any) -> None:
        self._get().debug(msg, *args, **kwargs)

    def info(self, msg: str, *args: Any, **kwargs: Any) -> None:
        self._get().info(msg, *args, **kwargs)

    def warning(self, msg: str, *args: Any, **kwargs: Any) -> None:
        self._get().warning(msg, *args, **kwargs)

    def error(self, msg: str, *args: Any, **kwargs: Any) -> None:
        self._get().error(msg, *args, **kwargs)


def get_logger(name: str) -> ChaosLogger:
    """Internal — used by all library modules.

    Returns a lazy proxy so that ``configure_chaos_logger`` takes effect
    even when called after module import.
    """
    return _LazyLogger(name)  # type: ignore[return-value]
